### PR TITLE
fix for issue 636

### PR DESCRIPTION
**Pull request by madhuy**
Resetting update_required flag after sending policy update.otherwise it will continiously calling policy update in case of ENABLE_EXTENDED_POLICY

### CODE REVIEW

File: src/components/application_manager/src/message_helper.cc [hunk 1]
```diff
@@ -1602,6 +1602,8 @@ void MessageHelper::SendPolicyUpdate(
     object[strings::msg_params][hmi_request::retry][i] = retries[i];
   }
   ApplicationManagerImpl::instance()->ManageHMICommand(message);
+  // to Reset the update_required_ flag
+  policy::PolicyHandler::instance()->OnUpdateRequestSentToMobile();
```
Comment (anosach-luxoft) [inline src/components/application_manager/src/message_helper.cc]: @madhuy I see that `OnUpdateRequestSentToMobile()` method performs a lot of logic but not only `update_required` flag resetting. Are you sure you need to start this logic here?

**Comment by madhuy** [comment]
@anosach-luxoft 

I am fallowing the flow from [code](https://github.com/smartdevicelink/sdl_core/blob/master/src/components/application_manager/src/application_manager_impl.cc#L1269) 

The above flow is used to update policy with "BasicCommunication.OnSystemRequest"
Same flow I am applying in case of "BasicCommunication. PolicyUpdate"

Hope fully it should be fine. Please do let me know if you have any concerns.

**Comment by madhuy** [comment]
@anosach-luxoft Could you please review this PR

**Comment by madhuy** [comment]
@anosach-luxoft do you any concern regarding this PR ?

**Comment by anosach-luxoft** [comment]
@madhuy sorry for delay, tomorrow I'll review and update you

**Comment by anosach-luxoft** [comment]
@madhuy I have only one concern regarding your implementation. I see that function `MessageHelper::SendPolicyUpdate()` sends message to HMI, but you're going to call `PolicyHandler::instance()->OnUpdateRequestSentToMobile()` which name looks like it should be called after sending message to mobile. As for me it looks strange. @AGaliuzov please help us to clarify.

**Comment by justinjdickow** [comment]
@AGaliuzov Looking for your input on this PR

**Comment by AGaliuzov** [comment]
@madhuy For mee seems like you have called the api for only reason - to reset the appropriate flag. Which is looks wrong. I'm agree with @anosach-luxoft . In case this flag should be changed we have to find proper way to do this.
So, for me it looks like we need to debug the whole sequence and find the real cause of the described problem

**Comment by madhuy** [comment]
@AGaliuzov we need to set 'update_required_' to 'false'. right now it is been set only from the onSystemRequest().  I think it is better to set only after policy update being completed. Please suggest the right place .

**Comment by madhuy** [comment]
Now restting update_required_ flag in UpdateStatusManager::OnValidUpdateReceived()`

Thanks @justin for your suggestion.

**Comment by LuxoftAKutsan** [comment]
@madhuy  do we have any unit tests for this logic. If not, is it possible to create? Does this changes fails existed unit tests?

**Comment by LuxoftAKutsan** [comment]
> @AGaliuzov we need to set 'update_required_' to 'false'. right now it is been set only from the onSystemRequest(). I think it is better to set only after policy update being completed. Please suggest the right place .

@madhuy you is right, but shouldn't we remove setting update_required to false from oSystemRequest?

**Comment by madhuy** [comment]
@JackLivio  Please comment on @LuxoftAKutsan review comments.

**Comment by justinjdickow** [comment]
@AGaliuzov Can you explain in more detail the case where this logic breaks a policy table update? Maybe there is a test case that covers the expectations for the update that @madhuy can work with?

**Comment by AGaliuzov** [comment]
@justinjdickow  I'm not sure about test cases but the flow is the next
1. PTU has been triggered but not yet finished
2. meanwhile user has requested another PTU directly from HMI - SDL set the flag `update_required` to handle this case.
3. First triggered PTU has been finished - SDL check if another update required and triggers on more PTU

Hence we don't miss the requested by user PTU. 
Let me know in case of any further questions

**Comment by AnastasiyaBritanova** [comment]
@madhuy , @AGaliuzov , @joeygrover 
1. if you review the [diagram](https://cloud.githubusercontent.com/assets/11644716/16690788/3d946dde-4532-11e6-976c-f4a20d7d5318.png) you'd see that SDL changes the flag `UPDATE_NEEDED` to `UPDATING` right after `OnSystemRequest` is sent to mobile app.
2. Since @madhuy's HMI **does not send** `OnSystemRequest` per [comment](https://github.com/smartdevicelink/sdl_core/issues/636#issuecomment-232197490), there is no trigger for SDL to reset this flag.
3. Resetting this flag after `OnReceivedPolicyUpdate` is a **wrong** solution, since there is a requirement for SDL to start another PTU if any trigger occurred and changed teh flag to `UPDATE_NEEDED`

**Suggested solution:**
Reset the flag `UPDATE_NEEDED` to `UPDATING` right after `BC.PolicyUpdate` with SnapshotPT is sent to HMI.
By pre-analysis, this should fix "continuous PTU" scenario.

**Comment by joeygrover** [comment]
@AnastasiyaBritanova's solution seems to be a more logical fix.

@madhuy can you comment?

**Comment by madhuy** [comment]
@joeygrover I don't have any issue if @AnastasiyaBritanova suggestion fixes the issue. 
@AnastasiyaBritanova Could please update this PR with your suggestions so that i can experiment?

**Comment by AnastasiyaBritanova** [comment]
@madhuy, do you mean this (from my previous [comment](https://github.com/smartdevicelink/sdl_core/pull/644#issuecomment-239119413) ): 
>Reset the flag UPDATE_NEEDED to UPDATING right after BC.PolicyUpdate with SnapshotPT is sent to HMI.

**Comment by madhuy** [comment]
@AnastasiyaBritanova Yes.

**Comment by madhuy** [comment]
@AnastasiyaBritanova Could you please provide me suggestion to fix this.

**Comment by AnastasiyaBritanova** [comment]
@madhuy 
please see the diagram attached - the marked section: SDL should reset status to `UPDATING` right after it sends `BC.PolicyUpdate` to HMI (instead of previously implemented when SDL changed to `UPDATING` only after `OnSystemRequest` was sent to the app).

proprietary_ptu_flow

**Comment by madhuy** [comment]
@AnastasiyaBritanova please Review #865 .

**Comment by AnastasiyaBritanova** [comment]
@AGaliuzov please review #865

**Comment by mghiumiusliu** [comment]
@madhuy do you plan to resolve conflicts in this PR?

**Comment by madhuy** [comment]
@mghiumiusliu  Closing this there s another PR for sam #865. Please review #865

**Pull request closed without merging**